src/System.Security.AccessControl/src/System/Security/AccessControl/CommonObjectSecurity.cs

```diff
@@ -0,0 +1,700 @@
+        {
+            if (rule == null)
+            {
+                throw new ArgumentNullException("rule");
+            }
+            Contract.EndContractBlock();
+
+            WriteLock();
+            try
+            {
+                bool result = true;
+
+                if (_securityDescriptor.DiscretionaryAcl == null)
+                {
+                    if (modification == AccessControlModification.Remove ||
+                        modification == AccessControlModification.RemoveAll ||
+                        modification == AccessControlModification.RemoveSpecific)
+                    {
+                        modified = false;
+                        return result;
+                    }
+
+                    _securityDescriptor.DiscretionaryAcl = new DiscretionaryAcl(IsContainer, IsDS, GenericAcl.AclRevision, 1);
+                    _securityDescriptor.AddControlFlags(ControlFlags.DiscretionaryAclPresent);
+                }
+
+                SecurityIdentifier sid = rule.IdentityReference.Translate(typeof(SecurityIdentifier)) as SecurityIdentifier;
+
+                if (rule.AccessControlType == AccessControlType.Allow)
+                {
+                    switch (modification)
+                    {
+                        case AccessControlModification.Add:
+                            _securityDescriptor.DiscretionaryAcl.AddAccess(AccessControlType.Allow, sid, rule.AccessMask, rule.InheritanceFlags, rule.PropagationFlags);
+                            break;
+
+                        case AccessControlModification.Set:
+                            _securityDescriptor.DiscretionaryAcl.SetAccess(AccessControlType.Allow, sid, rule.AccessMask, rule.InheritanceFlags, rule.PropagationFlags);
+                            break;
+
+                        case AccessControlModification.Reset:
+                            _securityDescriptor.DiscretionaryAcl.RemoveAccess(AccessControlType.Deny, sid, -1, InheritanceFlags.ContainerInherit | InheritanceFlags.ObjectInherit, 0);
+                            _securityDescriptor.DiscretionaryAcl.SetAccess(AccessControlType.Allow, sid, rule.AccessMask, rule.InheritanceFlags, rule.PropagationFlags);
+                            break;
+
+                        case AccessControlModification.Remove:
+                            result = _securityDescriptor.DiscretionaryAcl.RemoveAccess(AccessControlType.Allow, sid, rule.AccessMask, rule.InheritanceFlags, rule.PropagationFlags);
+                            break;
+
+                        case AccessControlModification.RemoveAll:
+                            result = _securityDescriptor.DiscretionaryAcl.RemoveAccess(AccessControlType.Allow, sid, -1, InheritanceFlags.ContainerInherit | InheritanceFlags.ObjectInherit, 0);
+                            if (result == false)
+                            {
+                                Contract.Assert(false, "Invalid operation");
```

Comment:
`Debug.Assert` should be used instead of `Contract.Assert` in CoreFX. Applies throughout.